apps/fhir/server/authentication.py

```diff
@@ -8,10 +9,111 @@
 logger = logging.getLogger('hhs_server.%s' % __name__)
 
 
-def match_hicn_hash(hicn_hash):
+def log_fhir_id_not_found(mbi_hash, hicn_hash,
+                          hash_lookup_type, hash_lookup_mesg):
+    '''
+        Logging for "FhirIDNotFound" type
+        used in match_backend_patient_identifier()
+    '''
+    logger.info(json.dumps({
+        "type": "FhirIDNotFound",
+        "mbi_hash": mbi_hash,
+        "hicn_hash": hicn_hash,
+        "hash_lookup_type": hash_lookup_type,
+        "hash_lookup_mesg": hash_lookup_mesg,
+    }))
+
+
+def log_fhir_id_is_found(fhir_id, mbi_hash, hicn_hash,
+                         hash_lookup_type, hash_lookup_mesg):
+    '''
+        Logging for "FhirIDFound" type
+        used in match_backend_patient_identifier()
+    '''
+    logger.info(json.dumps({
+        "type": "FhirIDFound",
+        "fhir_id": fhir_id,
+        "mbi_hash": mbi_hash,
+        "hicn_hash": hicn_hash,
+        "hash_lookup_type": hash_lookup_type,
+        "hash_lookup_mesg": hash_lookup_mesg,
+    }))
+
+
+def match_backend_patient_identifier(mbi_hash, hicn_hash):
+    '''
+    Matches a patient identifier via the backend FHIR server
+    using an MBI or HICN hash.
+
+    Summary:
+        1. mbi_hash is used for the initial lookup.
+        2. If there is a mbi lookup issue, the hicn_hash is used next.
+        3. If there is a hicn lookup issue, exceptions are raised.
+        4. A NotFound exception is raised if no match was found.
+
+    Returns:
+        fhir_id = Matched patient identifier.
+        backend_data = Passed thru to caller. Utilized in TestAuthentication.
+        hash_lookup_type = The type used for the successful lookup (M or H).
+
+    Raises:
+        UpstreamServerException: If hicn_hash search found duplicates.
+        NotFound: If both searches did not match a fhir_id.
+    '''
     auth_state = FhirServerAuth(None)
     certs = (auth_state['cert_file'], auth_state['key_file'])
-    # URL for patient ID.
+
+    # 1. mbi_hash is used for the initial lookup.
+    hash_lookup_type = "M"
+    hash_lookup_mesg = None
+
+    if mbi_hash is not None:
+        # URL for patient ID by mbi_hash.
+        url = get_resourcerouter().fhir_url + \
+            "Patient/?identifier=https%3A%2F%2Fbluebutton.cms.gov" + \
+            "%2Fresources%2Fidentifier%2Fmbi-hash%7C" + \
+            mbi_hash + \
+            "&_format=application%2Fjson%2Bfhir"
```

Comment:
@JFU-GIT Per your suggestion yesterday, I'll add some configurable settings in the base.py related to both the mbi_hash and hicn_hash search query parameters.

Reply:
that's great!